flow/activities/maintenance_activity.go

```diff
@@ -0,0 +1,298 @@
+package activities
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"go.temporal.io/sdk/activity"
+	"go.temporal.io/sdk/client"
+
+	"github.com/PeerDB-io/peer-flow/alerting"
+	"github.com/PeerDB-io/peer-flow/generated/protos"
+	"github.com/PeerDB-io/peer-flow/model"
+	"github.com/PeerDB-io/peer-flow/peerdbenv"
+	"github.com/PeerDB-io/peer-flow/shared"
+	"github.com/PeerDB-io/peer-flow/shared/telemetry"
+)
+
+const (
+	mirrorStateBackup   = "backup"
+	mirrorStateRestored = "restore"
+)
+
+type MaintenanceActivity struct {
+	CatalogPool    *pgxpool.Pool
+	Alerter        *alerting.Alerter
+	TemporalClient client.Client
+}
+
+type MaintenanceMirrorsInfo struct {
+	Mirrors []MaintenanceMirrorInfoItem
+}
+type MaintenanceMirrorInfoItem struct {
+	CreatedAt  time.Time
+	Name       string
+	WorkflowId string
+	Id         int64
+	IsCDC      bool
+}
+
+func (a *MaintenanceActivity) GetAllMirrors(ctx context.Context) (MaintenanceMirrorsInfo, error) {
+	rows, err := a.CatalogPool.Query(ctx,
+		`
+	select distinct on(f.name)
+	  f.id, f.name, f.workflow_id,
+	  f.created_at, coalesce(f.query_string, '')='' is_cdc
+	from flows f
+	`)
+	if err != nil {
+		return MaintenanceMirrorsInfo{}, err
+	}
+
+	maintenanceMirrorItems, err := pgx.CollectRows(rows, func(row pgx.CollectableRow) (MaintenanceMirrorInfoItem, error) {
+		var info MaintenanceMirrorInfoItem
+		err := row.Scan(&info.Id, &info.Name, &info.WorkflowId, &info.CreatedAt, &info.IsCDC)
+		return info, err
+	})
+	return MaintenanceMirrorsInfo{
+		Mirrors: maintenanceMirrorItems,
+	}, err
+}
+
+func (a *MaintenanceActivity) getMirrorStatus(ctx context.Context, mirrorInfo MaintenanceMirrorInfoItem) (protos.FlowStatus, error) {
+	encodedState, err := a.TemporalClient.QueryWorkflow(ctx, mirrorInfo.WorkflowId, "", shared.FlowStatusQuery)
+	if err != nil {
+		slog.Error("Error querying mirror status for maintenance", "mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "error", err)
+		return protos.FlowStatus_STATUS_UNKNOWN, err
+	}
+	var state protos.FlowStatus
+	err = encodedState.Get(&state)
+	if err != nil {
+		slog.Error("Error decoding mirror status for maintenance", "mirror", mirrorInfo.Name, "workflowId", mirrorInfo.WorkflowId, "error", err)
+		return protos.FlowStatus_STATUS_UNKNOWN,
+			fmt.Errorf("error decoding mirror status for maintenance: %w", err)
+	}
+	return state, nil
+}
+
+func (a *MaintenanceActivity) WaitForRunningSnapshots(ctx context.Context) (MaintenanceMirrorsInfo, error) {
+	mirrors, err := a.GetAllMirrors(ctx)
+	if err != nil {
+		return MaintenanceMirrorsInfo{}, err
+	}
+
+	slog.Info("Found mirrors for snapshot check", "mirrors", mirrors, "len", len(mirrors.Mirrors))
+
+	waitBeforeAlertingDuration := 30 * time.Minute
```

Comment:
a lot of this waiting logic could exist in workflow, using `workflow.Sleep` & `workflow.Go`

Reply:
I think I'll just remove it, `PeerDBMaintenanceModeWaitAlertSeconds` should be enough alerting for us, if anything is needed, we can just add it back in